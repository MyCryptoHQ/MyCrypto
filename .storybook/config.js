import React from 'react';
import { addDecorator } from '@storybook/react';
import { withA11y } from '@storybook/addon-a11y';
import { withThemesProvider } from 'storybook-addon-styled-component-theme';
import styled from 'styled-components';
import { light, dark } from '@mycrypto/ui';

import { GAU_COLORS } from 'v2/theme';

import 'sass/styles'

const LIGHT_THEME = {
  name: 'Light',
  ...light,
  GAU: {
    COLORS: GAU_COLORS
  }
};

const DARK_THEME = {
  name: 'Dark',
  ...dark,
  GAU: {
    COLORS: GAU_COLORS
  }
};

addDecorator(withThemesProvider([LIGHT_THEME, DARK_THEME]));
addDecorator(withA11y);
