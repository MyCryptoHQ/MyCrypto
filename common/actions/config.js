// @flow

/*** Change Language ***/
export type ChangeLanguageAction = {
  type: 'CONFIG_LANGUAGE_CHANGE',
  value: string
};

export function changeLanguage(sign: string): ChangeLanguageAction {
  return {
    type: 'CONFIG_LANGUAGE_CHANGE',
    value: sign
  };
}

/*** Change Node ***/
export type ChangeNodeAction = {
  type: 'CONFIG_NODE_CHANGE',
  // FIXME $keyof?
  value: string
};

export function changeNode(value: string): ChangeNodeAction {
  return {
    type: 'CONFIG_NODE_CHANGE',
    value
  };
}

/*** Change gas price ***/
export type ChangeGasPriceAction = {
  type: 'CONFIG_GAS_PRICE',
  value: number
};

export function changeGasPrice(value: number): ChangeGasPriceAction {
  return {
    type: 'CONFIG_GAS_PRICE',
    value
  };
}

/*** Union Type ***/
export type ConfigAction =
  | ChangeNodeAction
  | ChangeLanguageAction
  | ChangeGasPriceAction;
