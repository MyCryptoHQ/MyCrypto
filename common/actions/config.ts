/*** Change Language ***/
export interface ChangeLanguageAction {
  type: 'CONFIG_LANGUAGE_CHANGE';
  value: string;
}

export function changeLanguage(sign: string): ChangeLanguageAction {
  return {
    type: 'CONFIG_LANGUAGE_CHANGE',
    value: sign
  };
}

/*** Change Node ***/
export interface ChangeNodeAction {
  type: 'CONFIG_NODE_CHANGE';
  // FIXME $keyof?
  value: string;
}

export function changeNode(value: string): ChangeNodeAction {
  return {
    type: 'CONFIG_NODE_CHANGE',
    value
  };
}

/*** Change gas price ***/
export interface ChangeGasPriceAction {
  type: 'CONFIG_GAS_PRICE';
  value: number;
}

export function changeGasPrice(value: number): ChangeGasPriceAction {
  return {
    type: 'CONFIG_GAS_PRICE',
    value
  };
}

/*** Union Type ***/
export type ConfigAction =
  | ChangeNodeAction
  | ChangeLanguageAction
  | ChangeGasPriceAction;
