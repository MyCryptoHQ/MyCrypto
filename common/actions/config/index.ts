export * from './actionCreators';
export * from './actionTypes';
