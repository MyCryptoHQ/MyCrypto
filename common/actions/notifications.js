// @flow

/*** Shared types ***/
export type NOTIFICATION_LEVEL = 'danger' | 'warning' | 'success' | 'info';
export type INFINITY = 'infinity';

export type Notification = {
  level: NOTIFICATION_LEVEL,
  msg: string,
  duration?: number | INFINITY
};

/*** Show Notification ***/
export type ShowNotificationAction = {
  type: 'SHOW_NOTIFICATION',
  payload: Notification
};

export function showNotification(
  level: NOTIFICATION_LEVEL = 'info',
  msg: string,
  duration?: number
): ShowNotificationAction {
  return {
    type: 'SHOW_NOTIFICATION',
    payload: {
      level,
      msg,
      duration
    }
  };
}

/*** Close notification ***/
export type CloseNotificationAction = {
  type: 'CLOSE_NOTIFICATION',
  payload: Notification
};

export function closeNotification(
  notification: Notification
): CloseNotificationAction {
  return {
    type: 'CLOSE_NOTIFICATION',
    payload: notification
  };
}

/*** Union Type ***/
export type NotificationsAction =
  | ShowNotificationAction
  | CloseNotificationAction;
