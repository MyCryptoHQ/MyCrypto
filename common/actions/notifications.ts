import { ReactElement } from 'react';

/*** Shared types ***/
export type NOTIFICATION_LEVEL = 'danger' | 'warning' | 'success' | 'info';
export type INFINITY = 'infinity';

export interface Notification {
  level: NOTIFICATION_LEVEL;
  msg: ReactElement<any> | string;
  duration?: number | INFINITY;
}

/*** Show Notification ***/
export interface ShowNotificationAction {
  type: 'SHOW_NOTIFICATION';
  payload: Notification;
}

export function showNotification(
  level: NOTIFICATION_LEVEL = 'info',
  msg: ReactElement<any> | string,
  duration?: number | INFINITY
): ShowNotificationAction {
  return {
    type: 'SHOW_NOTIFICATION',
    payload: {
      level,
      msg,
      duration
    }
  };
}

/*** Close notification ***/
export interface CloseNotificationAction {
  type: 'CLOSE_NOTIFICATION';
  payload: Notification;
}

export function closeNotification(
  notification: Notification
): CloseNotificationAction {
  return {
    type: 'CLOSE_NOTIFICATION',
    payload: notification
  };
}

/*** Union Type ***/
export type NotificationsAction =
  | ShowNotificationAction
  | CloseNotificationAction;
