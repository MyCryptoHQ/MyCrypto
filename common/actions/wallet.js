// @flow
import type { IWallet } from 'libs/wallet/IWallet';
import Big from 'bignumber.js';
import { Wei } from 'libs/units';

/*** Unlock Private Key ***/
export type PrivateKeyUnlockParams = {
  key: string,
  password: string
};

export type UnlockPrivateKeyAction = {
  type: 'WALLET_UNLOCK_PRIVATE_KEY',
  payload: PrivateKeyUnlockParams
};

export function unlockPrivateKey(
  value: PrivateKeyUnlockParams
): UnlockPrivateKeyAction {
  return {
    type: 'WALLET_UNLOCK_PRIVATE_KEY',
    payload: value
  };
}

/*** Unlock Keystore File ***/
export type KeystoreUnlockParams = {
  file: string,
  password: string
};

export type UnlockKeystoreAction = {
  type: 'WALLET_UNLOCK_KEYSTORE',
  payload: KeystoreUnlockParams
};

export function unlockKeystore(
  value: KeystoreUnlockParams
): UnlockKeystoreAction {
  return {
    type: 'WALLET_UNLOCK_KEYSTORE',
    payload: value
  };
}

/*** Unlock Mnemonic ***/
export type MnemonicUnlockParams = {
  phrase: string,
  pass: string,
  path: string,
  address: string
};

export type UnlockMnemonicAction = {
  type: 'WALLET_UNLOCK_MNEMONIC',
  payload: MnemonicUnlockParams
};

export function unlockMnemonic(
  value: MnemonicUnlockParams
): UnlockMnemonicAction {
  return {
    type: 'WALLET_UNLOCK_MNEMONIC',
    payload: value
  };
}

/*** Set Wallet ***/
export type SetWalletAction = {
  type: 'WALLET_SET',
  payload: IWallet
};

export function setWallet(value: IWallet): SetWalletAction {
  return {
    type: 'WALLET_SET',
    payload: value
  };
}

/*** Set Balance ***/
export type SetBalanceAction = {
  type: 'WALLET_SET_BALANCE',
  payload: Wei
};

export function setBalance(value: Wei): SetBalanceAction {
  return {
    type: 'WALLET_SET_BALANCE',
    payload: value
  };
}

/*** Set Token Balance ***/
export type SetTokenBalancesAction = {
  type: 'WALLET_SET_TOKEN_BALANCES',
  payload: {
    [string]: Big
  }
};

export function setTokenBalances(payload: {
  [string]: Big
}): SetTokenBalancesAction {
  return {
    type: 'WALLET_SET_TOKEN_BALANCES',
    payload
  };
}

/*** Broadcast Tx ***/
export type BroadcastTxRequestedAction = {
  type: 'WALLET_BROADCAST_TX_REQUESTED',
  payload: {
    signedTx: string
  }
};

export function broadcastTx(signedTx: string): BroadcastTxRequestedAction {
  return {
    type: 'WALLET_BROADCAST_TX_REQUESTED',
    payload: {
      signedTx
    }
  };
}

/*** Union Type ***/
export type WalletAction =
  | UnlockPrivateKeyAction
  | SetWalletAction
  | SetBalanceAction
  | SetTokenBalancesAction
  | BroadcastTxRequestedAction;
