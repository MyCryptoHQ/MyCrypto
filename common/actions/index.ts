import { ConfigAction } from './config';
import { CustomTokenAction } from './customTokens';
import { DeterministicWalletAction } from './deterministicWallets';
import { EnsAction } from './ens';
import { NotificationsAction } from './notifications';
import { RatesAction } from './rates';
import { SwapAction } from './swap';
import { TransactionAction } from './transaction';
import { WalletAction } from './wallet';
export type AllActions =
  | ConfigAction
  | CustomTokenAction
  | DeterministicWalletAction
  | EnsAction
  | NotificationsAction
  | RatesAction
  | SwapAction
  | TransactionAction
  | WalletAction;
