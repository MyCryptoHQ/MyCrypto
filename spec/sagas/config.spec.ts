import { configuredStore } from 'store';
import { delay } from 'redux-saga';
import { call, cancel, fork, put, take, select } from 'redux-saga/effects';
import { cloneableGenerator, createMockTask } from 'redux-saga/utils';
import { toggleOfflineConfig, changeNode, changeNodeIntent, setLatestBlock } from 'actions/config';
import {
  pollOfflineStatus,
  handlePollOfflineStatus,
  handleNodeChangeIntent,
  unsetWeb3Node,
  unsetWeb3NodeOnWalletEvent,
  equivalentNodeOrDefault
} from 'sagas/config';
import { NODES, NodeConfig, NETWORKS } from 'config/data';
import {
  getNode,
  getNodeConfig,
  getOffline,
  getCustomNodeConfigs,
  getCustomNetworkConfigs
} from 'selectors/config';
import { INITIAL_STATE as configInitialState } from 'reducers/config';
import { getWalletInst } from 'selectors/wallet';
import { Web3Wallet } from 'libs/wallet';
import { RPCNode } from 'libs/nodes';
import { showNotification } from 'actions/notifications';
import { translateRaw } from 'translations';
import { resetWallet } from 'actions/wallet';
import { reset as resetTransaction } from 'actions/transaction';
// init module
configuredStore.getState();

describe('pollOfflineStatus*', () => {
  const nav = navigator as any;
  const doc = document as any;
  const data = {} as any;
  data.gen = cloneableGenerator(pollOfflineStatus)();
  const node = {
    lib: {
      ping: jest.fn()
    }
  };
  const isOffline = true;
  const raceSuccess = {
    pingSucceeded: true,
    timeout: false
  };
  const raceFailure = {
    pingSucceeded: false,
    timeout: true
  };

  let originalHidden;
  let originalOnLine;
  let originalRandom;

  beforeAll(() => {
    // backup global config
    originalHidden = document.hidden;
    originalOnLine = navigator.onLine;
    originalRandom = Math.random;

    // mock config
    Object.defineProperty(document, 'hidden', { value: false, writable: true });
    Object.defineProperty(navigator, 'onLine', { value: true, writable: true });
    Math.random = () => 0.001;
  });

  afterAll(() => {
    // restore global config
    Object.defineProperty(document, 'hidden', {
      value: originalHidden,
      writable: false
    });
    Object.defineProperty(navigator, 'onLine', {
      value: originalOnLine,
      writable: false
    });
    Math.random = originalRandom;
  });

  it('should select getNodeConfig', () => {
    expect(data.gen.next().value).toEqual(select(getNodeConfig));
  });

  it('should select getOffline', () => {
    expect(data.gen.next(node).value).toEqual(select(getOffline));
  });

  it('should call delay if document is hidden', () => {
    data.hiddenDoc = data.gen.clone();
    doc.hidden = true;
    expect(data.hiddenDoc.next(!isOffline).value).toEqual(call(delay, 1000));
    doc.hidden = false;
  });

  it('should race pingSucceeded and timeout', () => {
    data.isOfflineClone = data.gen.clone();
    data.shouldDelayClone = data.gen.clone();
    expect(data.gen.next(isOffline).value).toMatchSnapshot();
  });

  it('should toggle offline and show notification if navigator disagrees with isOffline and ping succeeds', () => {
    expect(data.gen.next(raceSuccess).value).toEqual(
      put(showNotification('success', 'Your connection to the network has been restored!', 3000))
    );
    expect(data.gen.next().value).toEqual(put(toggleOfflineConfig()));
  });

  it('should toggle offline and show notification if navigator agrees with isOffline and ping fails', () => {
    nav.onLine = isOffline;
    expect(data.isOfflineClone.next(!isOffline));
    expect(data.isOfflineClone.next(raceFailure).value).toMatchSnapshot();
    expect(data.isOfflineClone.next().value).toEqual(put(toggleOfflineConfig()));
    nav.onLine = !isOffline;
  });
});

describe('handlePollOfflineStatus*', () => {
  const gen = handlePollOfflineStatus();
  const mockTask = createMockTask();

  it('should fork pollOffineStatus', () => {
    const expectedForkYield = fork(pollOfflineStatus);
    expect(gen.next().value).toEqual(expectedForkYield);
  });

  it('should take CONFIG_STOP_POLL_OFFLINE_STATE', () => {
    expect(gen.next(mockTask).value).toEqual(take('CONFIG_STOP_POLL_OFFLINE_STATE'));
  });

  it('should cancel pollOfflineStatus', () => {
    expect(gen.next().value).toEqual(cancel(mockTask));
  });
});

describe('handleNodeChangeIntent*', () => {
  let originalRandom;

  // normal operation variables
  const defaultNode = configInitialState.nodeSelection;
  const defaultNodeConfig = NODES[defaultNode];
  const customNetworkConfigs = [];
  const defaultNodeNetwork = NETWORKS[defaultNodeConfig.network];
  const newNode = Object.keys(NODES).reduce(
    (acc, cur) => (NODES[acc].network === defaultNodeConfig.network ? cur : acc)
  );
  const newNodeConfig = NODES[newNode];
  const newNodeNetwork = NETWORKS[newNodeConfig.network];
  const changeNodeIntentAction = changeNodeIntent(newNode);
  const truthyWallet = true;
  const latestBlock = '0xa';
  const raceSuccess = {
    lb: latestBlock
  };
  const raceFailure = {
    to: true
  };

  const data = {} as any;
  data.gen = cloneableGenerator(handleNodeChangeIntent)(changeNodeIntentAction);

  function shouldBailOut(gen, nextVal, errMsg) {
    expect(gen.next(nextVal).value).toEqual(put(showNotification('danger', errMsg, 5000)));
    expect(gen.next().value).toEqual(
      put(changeNode(defaultNode, defaultNodeConfig, defaultNodeNetwork))
    );
    expect(gen.next().done).toEqual(true);
  }

  beforeAll(() => {
    originalRandom = Math.random;
    Math.random = () => 0.001;
  });

  afterAll(() => {
    Math.random = originalRandom;
  });

  it('should select getNode', () => {
    expect(data.gen.next().value).toEqual(select(getNode));
  });

  it('should select nodeConfig', () => {
    expect(data.gen.next(defaultNode).value).toEqual(select(getNodeConfig));
  });

  it('should select getCustomNetworkConfigs', () => {
    expect(data.gen.next(defaultNodeConfig).value).toEqual(select(getCustomNetworkConfigs));
  });

  it('should race getCurrentBlock and delay', () => {
    expect(data.gen.next(customNetworkConfigs).value).toMatchSnapshot();
  });

  it('should show error and revert to previous node if check times out', () => {
    data.clone1 = data.gen.clone();
    shouldBailOut(data.clone1, raceFailure, translateRaw('ERROR_32'));
  });

  it('should put setLatestBlock', () => {
    expect(data.gen.next(raceSuccess).value).toEqual(put(setLatestBlock(latestBlock)));
  });

  it('should put changeNode', () => {
    expect(data.gen.next().value).toEqual(
      put(changeNode(changeNodeIntentAction.payload, newNodeConfig, newNodeNetwork))
    );
  });

  it('should select getWalletInst', () => {
    expect(data.gen.next().value).toEqual(select(getWalletInst));
  });

  it('should call reload if wallet exists and network is new', () => {
    data.clone2 = data.gen.clone();
    expect(data.clone2.next(truthyWallet).value).toEqual(put(resetWallet()));
    expect(data.clone2.next(truthyWallet).value).toEqual(put(resetTransaction()));
    expect(data.clone2.next().done).toEqual(true);
  });

  it('should be done', () => {
    expect(data.gen.next().done).toEqual(true);
  });

  // custom node variables
  const customNodeConfigs = [
    {
      name: 'name',
      url: 'url',
      port: 443,
      network: 'network'
    }
  ];
  const customNodeIdFound = 'url:443';
  const customNodeIdNotFound = 'notFound';
  const customNodeAction = changeNodeIntent(customNodeIdFound);
  const customNodeNotFoundAction = changeNodeIntent(customNodeIdNotFound);
  data.customNode = handleNodeChangeIntent(customNodeAction);
  data.customNodeNotFound = handleNodeChangeIntent(customNodeNotFoundAction);

  // test custom node
  it('should select getCustomNodeConfig and match race snapshot', () => {
    data.customNode.next();
    data.customNode.next(defaultNode);
    data.customNode.next(defaultNodeConfig);
    expect(data.customNode.next(customNetworkConfigs).value).toEqual(select(getCustomNodeConfigs));
    expect(data.customNode.next(customNodeConfigs).value).toMatchSnapshot();
  });

  // test custom node not found
  it('should handle unknown / missing custom node', () => {
    data.customNodeNotFound.next();
    data.customNodeNotFound.next(defaultNode);
    data.customNodeNotFound.next(defaultNodeConfig);
    expect(data.customNodeNotFound.next(customNetworkConfigs).value).toEqual(
      select(getCustomNodeConfigs)
    );
    shouldBailOut(
      data.customNodeNotFound,
      customNodeConfigs,
      `Attempted to switch to unknown node '${customNodeNotFoundAction.payload}'`
    );
  });
});

describe('unsetWeb3Node*', () => {
  const node = 'web3';
  const mockNodeConfig = { network: 'ETH' } as any;
  const newNode = equivalentNodeOrDefault(mockNodeConfig);
  const gen = unsetWeb3Node();

  it('should select getNode', () => {
    expect(gen.next().value).toEqual(select(getNode));
  });

  it('should select getNodeConfig', () => {
    expect(gen.next(node).value).toEqual(select(getNodeConfig));
  });

  it('should put changeNodeIntent', () => {
    expect(gen.next(mockNodeConfig).value).toEqual(put(changeNodeIntent(newNode)));
  });

  it('should be done', () => {
    expect(gen.next().done).toEqual(true);
  });

  it('should return early if node type is not web3', () => {
    const gen1 = unsetWeb3Node();
    gen1.next();
    gen1.next('notWeb3');
    expect(gen1.next().done).toEqual(true);
  });
});

describe('unsetWeb3NodeOnWalletEvent*', () => {
  const fakeAction = {};
  const mockNode = 'web3';
  const mockNodeConfig: Partial<NodeConfig> = { network: 'ETH' };
  const gen = unsetWeb3NodeOnWalletEvent(fakeAction);

  it('should select getNode', () => {
    expect(gen.next().value).toEqual(select(getNode));
  });

  it('should select getNodeConfig', () => {
    expect(gen.next(mockNode).value).toEqual(select(getNodeConfig));
  });

  it('should put changeNodeIntent', () => {
    expect(gen.next(mockNodeConfig).value).toEqual(
      put(changeNodeIntent(equivalentNodeOrDefault(mockNodeConfig as any)))
    );
  });

  it('should be done', () => {
    expect(gen.next().done).toEqual(true);
  });

  it('should return early if node type is not web3', () => {
    const gen1 = unsetWeb3NodeOnWalletEvent({ payload: false });
    gen1.next(); //getNode
    gen1.next('notWeb3'); //getNodeConfig
    expect(gen1.next().done).toEqual(true);
  });

  it('should return early if wallet type is web3', () => {
    const mockAddress = '0x0';
    const mockNetwork = 'ETH';
    const mockWeb3Wallet = new Web3Wallet(mockAddress, mockNetwork);
    const gen2 = unsetWeb3NodeOnWalletEvent({ payload: mockWeb3Wallet });
    gen2.next(); //getNode
    gen2.next('web3'); //getNodeConfig
    expect(gen2.next().done).toEqual(true);
  });
});

describe('equivalentNodeOrDefault', () => {
  const originalNodeList = Object.keys(NODES);
  const appDefaultNode = configInitialState.nodeSelection;
  const mockNodeConfig = {
    network: 'ETH',
    service: 'fakeService',
    lib: new RPCNode('fakeEndpoint'),
    estimateGas: false
  };

  afterEach(() => {
    Object.keys(NODES).forEach(node => {
      if (originalNodeList.indexOf(node) === -1) {
        delete NODES[node];
      }
    });
  });

  it('should return node with equivalent network', () => {
    const node = equivalentNodeOrDefault({
      ...mockNodeConfig,
      network: 'Kovan'
    });
    expect(NODES[node].network).toEqual('Kovan');
  });

  it('should return app default if no eqivalent is found', () => {
    const node = equivalentNodeOrDefault({
      ...mockNodeConfig,
      network: 'noEqivalentExists'
    });
    expect(node).toEqual(appDefaultNode);
  });

  it('should ignore web3 from node list', () => {
    NODES.web3 = {
      ...mockNodeConfig,
      network: 'uniqueToWeb3'
    };

    const node = equivalentNodeOrDefault({
      ...mockNodeConfig,
      network: 'uniqueToWeb3'
    });
    expect(node).toEqual(appDefaultNode);
  });
});
