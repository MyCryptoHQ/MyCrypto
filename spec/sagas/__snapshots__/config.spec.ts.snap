// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleNodeChangeIntent* should race getCurrentBlock and delay 1`] = `
Object {
  "@@redux-saga/IO": true,
  "RACE": Object {
    "lb": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [],
        "context": null,
        "fn": [Function],
      },
    },
    "to": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          5000,
        ],
        "context": null,
        "fn": [Function],
      },
    },
  },
}
`;

exports[`handleNodeChangeIntent* should select getCustomNodeConfig and match race snapshot 1`] = `
Object {
  "@@redux-saga/IO": true,
  "RACE": Object {
    "lb": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [],
        "context": null,
        "fn": [Function],
      },
    },
    "to": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          5000,
        ],
        "context": null,
        "fn": [Function],
      },
    },
  },
}
`;

exports[`pollOfflineStatus* should race pingSucceeded and timeout 1`] = `
Object {
  "@@redux-saga/IO": true,
  "RACE": Object {
    "pingSucceeded": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [],
        "context": null,
        "fn": [Function],
      },
    },
    "timeout": Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          5000,
        ],
        "context": null,
        "fn": [Function],
      },
    },
  },
}
`;

exports[`pollOfflineStatus* should toggle offline and show notification if navigator agrees with isOffline and ping fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "duration": 5000,
        "id": 0.001,
        "level": "info",
        "msg": "You are currently offline. Some features will be unavailable.",
      },
      "type": "SHOW_NOTIFICATION",
    },
    "channel": null,
  },
}
`;
